compiler/src/iree/compiler/GlobalOptimization/FuseDequantizationMatmul.cpp

```diff
@@ -0,0 +1,860 @@
+                                                     regionOp);
+  if (failed(maybeFusedRegionOp))
+    return failure();
+
+  if (fill && *fill) {
+    FailureOr<IREE::Flow::DispatchRegionOp> maybeFusedFillRegionOp =
+        IREE::Flow::clonePrecedingOpIntoDispatchRegion(rewriter, fill.value(),
+                                                       regionOp);
+    if (failed(maybeFusedFillRegionOp))
+      return failure();
+  }
+
+  return success();
+}
+
+static FailureOr<IREE::Flow::DispatchRegionOp>
+wrapConsecutiveOpsInDispatchRegion(RewriterBase &rewriter,
+                                   SmallVector<Operation *> ops) {
+  FailureOr<IREE::Flow::DispatchRegionOp> maybeRegionOp =
+      IREE::Flow::wrapOpInDispatchRegion(rewriter, ops.back());
+  if (failed(maybeRegionOp)) {
+    return failure();
+  }
+  IREE::Flow::DispatchRegionOp regionOp = maybeRegionOp.value();
+
+  SmallVector<Operation *> precedingOps(ops.begin(), ops.end() - 1);
+  FailureOr<IREE::Flow::DispatchRegionOp> maybeFusedRegionOp =
+      IREE::Flow::movePrecedingOpsIntoDispatchRegion(rewriter, precedingOps,
+                                                     regionOp);
+  if (failed(maybeFusedRegionOp)) {
+    return failure();
+  }
+  regionOp = maybeFusedRegionOp.value();
+
+  return regionOp;
+}
+
+static SmallVector<utils::IteratorType>
+getParallelAndReductionIterators(unsigned nLoops, unsigned nReduction) {
+  SmallVector<utils::IteratorType> res(nLoops - nReduction,
+                                       utils::IteratorType::parallel);
+  res.append(nReduction, utils::IteratorType::reduction);
+  return res;
+}
+
+// Takes as input the dequantization `linalg.generic` op and the matmul
+// `linalg.generic` op, and returns the scales, zero points, quantized
+// input matrix, unquantizaed input matrix, and dequantized result
+// matrix.
+static std::optional<SmallVector<OpOperand *>>
+getDequantMatmulInputs_f32(linalg::GenericOp dequant,
+                           linalg::GenericOp matmul) {
+  OpOperand *scales, *zps, *quantMat, *unquantMat, *dequantMat;
+  for (int operandIdx = 0; operandIdx < dequant.getNumDpsInputs();
+       operandIdx++) {
+    OpOperand *operand = dequant.getDpsInputOperand(operandIdx);
+    Value input = operand->get();
+    RankedTensorType inputType =
+        llvm::dyn_cast<RankedTensorType>(input.getType());
+    if (!inputType) {
+      continue;
+    }
+    if (inputType.getElementTypeBitWidth() != 32) {
+      quantMat = operand;
+      continue;
+    }
+    for (Operation &bodyOp : dequant.getBlock()->getOperations()) {
+      if (isa<arith::MulFOp>(bodyOp)) {
+        if (bodyOp.getOperand(1) ==
+            dequant.getBlock()->getArgument(operandIdx)) {
+          scales = operand;
+          break;
+        }
+      } else if (isa<arith::SubFOp>(bodyOp)) {
+        if (bodyOp.getOperand(1) ==
+            dequant.getBlock()->getArgument(operandIdx)) {
+          zps = operand;
+          break;
+        }
```

Comment:
Are we OK with the loose matching semantics here or do we want to ensure that we only do rewrites that are technically correct?  If we are OK with the loose matching semantics, what's the rationale for why we don't think that false-matches are going to be a problem --- is this pass going to stay behind a flag and not enabled by default ?
@MaheshRavishankar

Reply:
Those are  the semantics that we've used for the simple fusion of the dequantization and matmul ops into a dispatch region, so it seemed reasonable in that case. I agree it would make sense to have a stricter matching now that we are doing a big rewrite, but for now I think I can leave it as a TODO comment since it is all behind a flag anyway

Reply:
Also, I'm realizing now that this blob is actually from the helper function that gets the dequantization and matmul inputs. So the matching on the ops actually happens before. The operations are known to be a dequant and matmul at this point, so this matching is only to find which operand is the scales and which is the zero points.

Reply:
Maybe add an assert at the start of the function that checks that this is a grouped dequantization ops that ensures that it has the expected form.